### PR TITLE
Mark fields that have a 'skip_serializing_if' annotation as optional during TS generation

### PR DESCRIPTION
**Pull request by sagacity**

### CODE REVIEW

File: examples/example-deno-runtime/tests.ts [hunk 1]
```diff
@@ -295,6 +294,11 @@ const imports: Imports = {
       },
     });
   },
+
+  importStructWithOptions: (arg: StructWithOptions): StructWithOptions => {
+    assertEquals(arg.potentiallyOptionalString, "Hello");
```
Comment (arendjr) [inline examples/example-deno-runtime/tests.ts:299]: I think we should also add a test case for the omitted value, so we verify whether it's `null` or `undefined`.

File: examples/example-protocol/src/assets/rust_wasmer_runtime_test/expected_types.rs [hunk 2]
```diff
@@ -256,3 +256,10 @@ pub struct StructWithGenerics<T> {
     #[serde(default, skip_serializing_if = "Option::is_none")]
     pub optional_timestamp: Option<MyDateTime>,
 }
+
+#[derive(Clone, Debug, Deserialize, PartialEq, Serialize)]
+#[serde(rename_all = "camelCase")]
+pub struct StructWithOptions {
+    #[serde(default, skip_serializing_if = "String::is_empty")]
+    pub potentially_optional_string: String,
```
Comment (arendjr) [inline examples/example-protocol/src/assets/rust_wasmer_runtime_test/expected_types.rs:264]: I think this also needs one additional case: An `Option` value without the `skip_serializing_if` annotation. If I'm not mistaken, the currently generated TS types will assume that to be `undefined` too, though it would be serialized as `null`.

File: fp-bindgen/src/generators/ts_runtime/mod.rs [hunk 3]
```diff
@@ -905,9 +905,10 @@ fn format_struct_fields(fields: &[Field], types: &TypeMap, casing: Casing) -> Ve
     fields
         .iter()
         .flat_map(|field| {
+            let field_optional = field.attrs.skip_serializing_if.is_some();
             let field_decl = match types.get(&field.ty) {
                 Some(Type::Container(name, _)) => {
-                    let optional = if name == "Option" { "?" } else { "" };
+                    let field_optional = field_optional || name == "Option";
```
Comment (arendjr) [inline fp-bindgen/src/generators/ts_runtime/mod.rs:911]: This is a bit tricky now. We could still auto opt-in to skipping the serialization for `Option`s, but then only if the `#[fp(rust_module = "...")]` annotation is *not* used. But that would become rather hard to explain, so I would suggest we just completely drop doing it automatically and leave the decision entirely to the user. Just as Serde does, basically.

It does warrant some explanation of the implications in the CHANGELOG though, which is probably a good moment to pick up FP-2725 right away and use the https://keepachangelog.com/en/1.0.0/ convention here too.

File: examples/example-protocol/src/assets/rust_wasmer_runtime_test/expected_types.rs (hunk 2 above)
Comment (sagacity) [reply examples/example-protocol/src/assets/rust_wasmer_runtime_test/expected_types.rs:264]: I'll extend the struct with a bunch of different fields (some filled, some empty).

File: fp-bindgen/src/generators/ts_runtime/mod.rs (hunk 3 above)
Comment (sagacity) [reply fp-bindgen/src/generators/ts_runtime/mod.rs:911]: Isn't this behaviour what we want though? If it _could be potentially optional_ we just mark it as an optional field. This might be a bit too conservative in some cases but it's not that complicated to reason about, I think?

Comment (arendjr) [reply fp-bindgen/src/generators/ts_runtime/mod.rs:911]: Right, the difference is in whether the optional field should be serialized or not. If a `None` field is serialized it will be explicitly serialized as a `null` value and end up as such on the TypeScript side. If it's not serialized at all, it doesn't go over the wire and thus remains `undefined` on the TypeScript side. So the difference is between whether we want to generate `fieldName: T | null` or `fieldName?: T` in TypeScript.

Thus far, we have always assumed that when you specify `Option<T>` in your protocol, you would be happy to omit the `None` during serialization, and so we would generate `fieldName?: T` for TypeScript. We even enforced this when generating Rust models, by adding the `skip_serializing_if = "Option::is_none"` Serde annotation ourselves.

The problem, where this started to break down, is when Rust models were combined with the `rust_module` annotation. In that case we don't generate the Rust models, and thus we cannot enforce anything. We assumed the user would add the Serde annotation for consistency, but it's just an assumption... they might forget, or they might not even always want that.

So instead of assuming that the presence of an `Option` implies the usage of the `skip_serializing_if` annotation, we should just decouple the two: `Option` without the annotation should generate `fieldName: T | null` as type, while `Option` with the annotation should generate `fieldName?: T` (and also not `fieldName?: T | null`!).

Comment (sagacity) [reply fp-bindgen/src/generators/ts_runtime/mod.rs]: @arendjr Do the latest commit (add1fa4957ccf6fd2acc8d7603e503a9c619a7d6) make sense? The non-deno tests are failing since I didn't fix them yet, but do the generated types look like you'd expect?

For reference, the generated `StructWithOptions` now looks like this:
```
export type StructWithOptions = {
    filledString: string | null;
    emptyString: string | null;
    filledRegularOptionString: string | null;
    emptyRegularOptionString: string | null;
    undefinedRegularOptionString?: string;
};
```

Based off of:
```
#[derive(Clone, Debug, Default, Deserialize, PartialEq, Eq, Serialize, Serializable)]
#[fp(rename_all = "camelCase")]
pub struct StructWithOptions {
    #[serde(default, skip_serializing_if = "String::is_empty")]
    pub filled_string: String,
    #[serde(default, skip_serializing_if = "String::is_empty")]
    pub empty_string: String,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub filled_regular_option_string: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub empty_regular_option_string: Option<String>,
    #[serde(default)]
    pub undefined_regular_option_string: Option<String>,
}
```

Comment (arendjr) [reply fp-bindgen/src/generators/ts_runtime/mod.rs]: The TS types from your last comment are exactly opposite what I would expect :sweat_smile:

Comment (sagacity) [reply fp-bindgen/src/generators/ts_runtime/mod.rs]: 🤦 seems you're right

File: examples/example-protocol/src/assets/rust_plugin_test/expected_types.rs [hunk 4]
```diff
@@ -261,5 +261,15 @@ pub struct StructWithGenerics<T> {
...
+    pub filled_string: String,
+    #[serde(default, skip_serializing_if = "String::is_empty")]
+    pub empty_string: String,
+    #[serde(default, skip_serializing_if = "Option::is_none")]
+    pub filled_option_string: Option<String>,
+    #[serde(default, skip_serializing_if = "Option::is_none")]
+    pub empty_option_string: Option<String>,
+    #[serde(default, skip_serializing_if = "Option::is_none")]
+    pub never_skipped_filled_option_string: Option<String>,
+    #[serde(default, skip_serializing_if = "Option::is_none")]
```
Comment (arendjr) [inline examples/example-protocol/src/assets/rust_plugin_test/expected_types.rs:273]: Why do these "never skipped" fields still have the `skip` annotations?

Comment (sagacity) [reply examples/example-protocol/src/assets/rust_plugin_test/expected_types.rs:273]: This is the default behaviour for the Rust struct generator, it always adds this attribute whenever an `Option` is used if it's not there yet. The type defined in the actual protocol doesn't have these attributes.

Are you saying you'd want to change this behaviour, too? It seems like a decent default.

Comment (arendjr) [reply examples/example-protocol/src/assets/rust_plugin_test/expected_types.rs:273]: I would change them, yes, because otherwise the behavior is different between the situation when we generate a struct versus when we reuse an existing struct. If we skip the serialization, the resulting type could also never be `null`, it would always be `undefined`. So this led me to look at how the tests could pass.... I think we're making an error by using `assertEquals()` in the tests instead of `assertStrictEquals()`. I highly suspect as soon as we change the assertions, you'll see the values that are asserted to be `null` are really `undefined`.

**Review by arendjr — APPROVED** [review]
:+1:
